File: windows-player/src/HomeKtv.Windows/Playback/ActivePlayerLeaseGate.cs
namespace HomeKtv.Windows.Playback;

public sealed record PlayerAssignment(string Role, long Generation, long LeaseMs);

/** Fences local projection and upstream reports when this client is standby or its lease expired. */
public sealed class ActivePlayerLeaseGate
{
    private readonly Func<DateTimeOffset> now;
    private readonly object sync = new();
    private long generation;
    private DateTimeOffset expiresAt;
    private bool active;

    public ActivePlayerLeaseGate(Func<DateTimeOffset>? now = null)
    {
        this.now = now ?? (() => DateTimeOffset.UtcNow);
    }

    public void Apply(PlayerAssignment assignment)
    {
        lock (sync)
        {
            active = string.Equals(assignment.Role, "ACTIVE", StringComparison.OrdinalIgnoreCase)
                     && assignment.Generation > 0 && assignment.LeaseMs > 0;
            generation = active ? assignment.Generation : 0;
            expiresAt = active ? now().AddMilliseconds(assignment.LeaseMs) : DateTimeOffset.MinValue;
        }
    }

    public bool TryGetGeneration(out long value)
    {
        lock (sync)
        {
            if (active && now() < expiresAt)
            {
                value = generation;
                return true;
            }
            active = false;
            generation = 0;
            value = 0;
            return false;
        }
    }

    public void Disconnect()
    {
        lock (sync)
        {
            active = false;
            generation = 0;
            expiresAt = DateTimeOffset.MinValue;
        }
    }
}

File: android-tv/app/src/main/java/com/homektv/tv/net/ActivePlayerLeaseGate.kt
package com.homektv.tv.net

import android.os.SystemClock

/** Fences playback and upstream reports unless this client owns a live server lease. */
class ActivePlayerLeaseGate(
    private val nowMs: () -> Long = { SystemClock.elapsedRealtime() },
) {
    private var active = false
    private var generation = 0L
    private var expiresAtMs = 0L

    @Synchronized
    fun apply(role: String, assignedGeneration: Long, leaseMs: Long) {
        active = role.equals("ACTIVE", ignoreCase = true) && assignedGeneration > 0 && leaseMs > 0
        generation = if (active) assignedGeneration else 0L
        expiresAtMs = if (active) nowMs() + leaseMs else 0L
    }

    @Synchronized
    fun activeGeneration(): Long? {
        if (!active || nowMs() >= expiresAtMs) {
            active = false
            generation = 0L
            return null
        }
        return generation
    }

    @Synchronized
    fun disconnect() {
        active = false
        generation = 0L
        expiresAtMs = 0L
    }
}
